File: Project_Codes/Python Version/dimensionality_reduction.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import RobustScaler
from sklearn.decomposition import PCA
import umap

# ---------------------------------------
# DATA PREPARATION
# ---------------------------------------
def prepare_data_for_dr(df):
    "Prepares, samples and scales the data for dimensionality reduction"

    # Take a sample to avoid memory and performance issues
    df_sample = df.sample(n=10000, random_state=42).copy()

    # Create text labels for the legend
    df_sample['Flight Status'] = np.where(df_sample['ARR_DELAY'] > 15, 'Delayed (>15m)', 'On-Time')

    # Select the columns to scale
    cols_to_scale = [
        'CRS_DEP_TIME', 'CRS_ARR_TIME', 'DISTANCE', 'CRS_ELAPSED_TIME', 'MONTH',
        'SCHEDULED_SPEED',
        'ORIGIN_CONGESTION',
        'DESTINATION_CONGESTION',
        'ROUTE_BOTTLENECK_INTERACTION',
        'ROUTE_FREQUENCY',
        'AIRLINE_HUB_DOMINANCE'
    ]

    # Create a copy for PCA/UMAP
    df_features = df_sample.copy()

    # Use RobustScaler to handle extreme delay outliers
    from sklearn.preprocessing import RobustScaler
    df = RobustScaler().fit_transform(df_features[cols_to_scale])
    x_scaled = RobustScaler().fit_transform(df_features[cols_to_scale])

    return x_scaled, df_features, cols_to_scale

# ---------------------------------------
# PCA
# ---------------------------------------
def run_and_plot_pca(x_scaled, df_features, cols_to_scale):
    """Runs PCA and generates the variance, projection, and loading plots."""

    # run PCA
    pca = PCA()

    # Fit the PCA model and compress the scaled data into principal components
    pca_result = pca.fit_transform(x_scaled)

    # Plot 1: Cumulative Explained Variance
    plt.figure(figsize=(8, 4))
    plt.plot(range(1, len(pca.explained_variance_ratio_) + 1),
             np.cumsum(pca.explained_variance_ratio_),
             marker='o', linestyle='-', color='dodgerblue')
    plt.axhline(y=0.80, color='r', linestyle='--', label='80% Information Retained')
    plt.title('Data Compression: How Much Flight Information is Preserved by PCA?', fontsize=14, pad=15)
    plt.xlabel('Number of Principal Components (Compressed Features)')
    plt.ylabel('Percentage of Original Flight Data Retained')
    plt.legend(loc='lower right', bbox_to_anchor=(0.95, 0.35))
    sns.despine()
    plt.show()
    

    # Plot 2: PCA 2D Projection
    plt.figure(figsize=(10, 6))
    sns.scatterplot(x=pca_result[:, 0], y=pca_result[:, 1],
                    hue=df_features['Flight Status'].values,
                    palette={'On-Time': 'lightseagreen', 'Delayed (>15m)': 'crimson'},
                    alpha=0.5, s=15, linewidth=0)
    plt.title('Linear Dimensionality Reduction: PCA (2D Projection)', fontsize=14, pad=15)
    plt.xlabel('Principal Component 1')
    plt.ylabel('Principal Component 2')
    plt.legend(title='Flight Outcome', loc='upper right')
    sns.despine()
    plt.show()
    

    # Plot 3: PCA Component Loadings (Heatmap)
    loadings = pd.DataFrame(
        pca.components_,
        columns=cols_to_scale,
        index=[f'PC{i + 1}' for i in range(pca.n_components_)]
    )

    plt.figure(figsize=(12, 6))
    sns.heatmap(loadings.head(3), annot=True, cmap='coolwarm', center=0, fmt=".2f")
    plt.title('PCA Component Loadings: Feature Contributions to Top 3 Components', fontsize=14, pad=15)
    plt.ylabel('Principal Components')
    plt.xlabel('Original Flight Features')
    plt.xticks(rotation=45, ha='right')  # Rotate x-labels to fit the feature names
    plt.tight_layout()
    plt.show()
    


# ---------------------------------------
# UMAP
# ---------------------------------------
def run_and_plot_umap(x_scaled, df_features):
    """Runs UMAP and generates the 2D projection plot."""

    # run UMAP
    reducer = umap.UMAP(
        n_neighbors=15,
        min_dist=0.1,
        n_components=2,
        random_state=42
    )

    # Fit the UMAP model to build a neighborhood graph and compress the data into a non-linear 2D map
    umap_result = reducer.fit_transform(x_scaled)

    # Plot 4: UMAP 2D Scatter Plot
    plt.figure(figsize=(10, 6))
    sns.scatterplot(x=umap_result[:, 0], y=umap_result[:, 1],
                    hue=df_features['Flight Status'].values,
                    palette={'On-Time': 'lightseagreen', 'Delayed (>15m)': 'crimson'},
                    alpha=0.5, s=15, linewidth=0)
    plt.title('Non-Linear Dimensionality Reduction: UMAP', fontsize=14, pad=15)
    plt.xlabel('UMAP Dimension 1')
    plt.ylabel('UMAP Dimension 2')
    plt.legend(title='Flight Outcome', loc='upper right')
    sns.despine()
    plt.show()
    


# ---------------------------------------
# DIMENSIONALITY REDUCTION PIPELINE
# ---------------------------------------
def execute_dimensionality_reduction(df):
    "Main execution function to run the entire DR pipeline"

    # 1. Data Preparation
    x_scaled, df_features, cols_to_scale = prepare_data_for_dr(df)

    # 2. PCA
    run_and_plot_pca(x_scaled, df_features, cols_to_scale)

    # 3. UMAP
    run_and_plot_umap(x_scaled, df_features)

File: Project_Codes/R Version/dimensionality_reduction.R
library(dplyr)
library(ggplot2)
library(umap)
library(reshape2)

### ---------------------------------------
### DATA PREPARATION
### ---------------------------------------
prepare_data_for_dr <- function(df) {
    # "Prepares, samples and scales the data for dimensionality reduction"

    # Safely sample the dataset (UMAP is computationally heavy on massive datasets)
    sample_size <- min(15000, nrow(df))
    set.seed(42)
    df_sample <- df[sample(nrow(df), sample_size), ]

    # Keep only numeric columns for DR algorithms
    df_features <- df_sample %>% select(where(is.numeric))
    cols_to_scale <- colnames(df_features)

    # Custom RobustScaler implementation (matches sklearn: median removal & IQR scaling)
    robust_scale <- function(x) {
        iqr_val <- IQR(x, na.rm = TRUE)
        if(iqr_val == 0) return(x - median(x, na.rm = TRUE)) # Avoid division by zero
        return((x - median(x, na.rm = TRUE)) / iqr_val)
    }

    x_scaled <- as.data.frame(lapply(df_features, robust_scale))

    return(list(x_scaled = x_scaled, df_features = df_features, cols_to_scale = cols_to_scale))
}

### ---------------------------------------
### PCA
### ---------------------------------------
run_and_plot_pca <- function(x_scaled, df_features, cols_to_scale) {
    # """Runs PCA and generates the variance, projection, and loading plots."""

    # Run PCA (data is already scaled via robust_scale)
    pca_result <- prcomp(x_scaled, center = FALSE, scale. = FALSE)

    # 1. Cumulative Variance Plot
    explained_variance <- pca_result$sdev^2 / sum(pca_result$sdev^2)
    cumulative_variance <- cumsum(explained_variance)

    var_df <- data.frame(PC = 1:length(cumulative_variance), Cumulative_Variance = cumulative_variance)

    p1 <- ggplot(var_df, aes(x = PC, y = Cumulative_Variance)) +
        geom_line(color = "dodgerblue", size = 1) +
        geom_point(color = "dodgerblue", size = 2) +
        geom_hline(yintercept = 0.80, color = "red", linetype = "dashed") +
        labs(title = 'PCA - Cumulative Explained Variance',
             x = 'Number of Principal Components', y = 'Cumulative Variance Retained') +
        theme_classic()
    print(p1)

    # 2. 2D Projection Plot
    proj_df <- data.frame(PC1 = pca_result$x[, 1], PC2 = pca_result$x[, 2])

    p2 <- ggplot(proj_df, aes(x = PC1, y = PC2)) +
        geom_point(alpha = 0.5, color = "steelblue") +
        labs(title = 'Linear Dimensionality Reduction: PCA (2D Projection)',
             x = 'Principal Component 1', y = 'Principal Component 2') +
        theme_classic()
    print(p2)

    # 3. PCA Component Loadings Heatmap (Top 3 Components)
    num_components <- min(3, ncol(pca_result$rotation))
    loadings <- pca_result$rotation[, 1:num_components]
    loadings_melted <- melt(loadings)
    colnames(loadings_melted) <- c("Feature", "PC", "Loading")

    p3 <- ggplot(loadings_melted, aes(x = PC, y = Feature, fill = Loading)) +
        geom_tile(color = "white") +
        scale_fill_gradient2(low = "firebrick", high = "dodgerblue", mid = "white", midpoint = 0) +
        labs(title = 'PCA - Component Loadings (Heatmap)') +
        theme_minimal() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
    print(p3)

    return(pca_result)
}

### ---------------------------------------
### UMAP
### ---------------------------------------
run_and_plot_umap <- function(x_scaled, df_features) {
    # """Runs UMAP and generates the 2D projection plot."""

    # Run UMAP
    set.seed(42)
    umap_result <- umap(x_scaled)

    # Projection Plot (2D)
    umap_df <- data.frame(UMAP1 = umap_result$layout[, 1], UMAP2 = umap_result$layout[, 2])

    p <- ggplot(umap_df, aes(x = UMAP1, y = UMAP2)) +
        geom_point(alpha = 0.5, color = "mediumseagreen") +
        labs(title = 'Non-Linear Dimensionality Reduction: UMAP',
             x = 'UMAP Dimension 1', y = 'UMAP Dimension 2') +
        theme_classic()
    print(p)

    return(umap_result)
}

### ---------------------------------------
### DIMENSIONALITY REDUCTION PIPELINE
### ---------------------------------------
execute_dimensionality_reduction <- function(df) {
    # "Main execution function to run the entire DR pipeline"

    cat("\n--- Starting Dimensionality Reduction Pipeline ---\n")

    # 1. Prepare data
    prep_results <- prepare_data_for_dr(df)

    # 2. PCA
    cat("Running PCA...\n")
    pca_result <- run_and_plot_pca(prep_results$x_scaled, prep_results$df_features, prep_results$cols_to_scale)

    # 3. UMAP
    cat("Running UMAP...\n")
    umap_result <- run_and_plot_umap(prep_results$x_scaled, prep_results$df_features)

    cat("--- Dimensionality Reduction Complete ---\n")
}
